pkg/apimachinery/patch/patch.go

```diff
@@ -38,9 +37,73 @@ const (
 	PatchRemoveOp  = "remove"
 )
 
+type PatchSet interface {
+	Test(path string, value interface{})
+	Add(path string, value interface{})
+	Remove(path string)
+	Replace(path string, value interface{})
+	TestAndReplace(path string, oldVal, newVal interface{})
+	AddOrReplace(path string, value interface{}, condition bool)
+	IsEmpty() bool
+	GeneratePayload() ([]byte, error)
+}
+
+type patchSet struct {
+	patches []PatchOperation
+}
+
+func New() PatchSet {
+	return &patchSet{}
+}
+
+func (p *patchSet) IsEmpty() bool {
+	return len(p.patches) < 1
+}
+
+func (p *patchSet) addOp(op, path string, value interface{}) {
+	p.patches = append(p.patches, PatchOperation{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	})
+}
+
+func (p *patchSet) Test(path string, value interface{}) {
+	p.addOp(PatchTestOp, path, value)
+}
+
+func (p *patchSet) Add(path string, value interface{}) {
+	p.addOp(PatchAddOp, path, value)
+}
+
+func (p *patchSet) Remove(path string) {
+	p.addOp(PatchRemoveOp, path, nil)
+}
+
+func (p *patchSet) Replace(path string, value interface{}) {
+	p.addOp(PatchReplaceOp, path, value)
+}
+
+func (p *patchSet) TestAndReplace(path string, oldVal, newVal interface{}) {
+	p.Test(path, oldVal)
+	p.Replace(path, newVal)
+}
+
+func (p *patchSet) AddOrReplace(path string, value interface{}, condition bool) {
```

Comment:
This is odd.
Why do we need a condition and not a direct usage?

Also, in most cases, `Replace` is not really needed, `Add` is enough and will do both (i.e. add if missing or override is already exists).

Reply:
Yes, I know in many places `replace` is used. My claim is that it is not needed.
There are also many places that `add`/`replace` is combined with `test` without a good reason.